lib/compat/wordpress-6.0/theme-json-webfonts-handler.php

```diff
@@ -0,0 +1,521 @@
+<?php
+/**
+ * Handler for theme.json webfonts.
+ *
+ * @since      6.0.0
+ * @subpackage WebFonts
+ * @package    WordPress
+ */
+
+if ( ! function_exists( '_wp_theme_json_webfonts_handler' ) ) {
+	add_action( 'plugins_loaded', '_wp_theme_json_webfonts_handler' );
+	/**
+	 * Runs the theme.json webfonts handler to expose defined webfonts
+	 * to the editor's typography pickers and generate '@font-face' style
+	 * declarations.
+	 *
+	 * This is not a public API, but rather an internal handler.
+	 * A future public Webfonts API will replace this code.
+	 *
+	 * This code design is intentional to encapsulate the inner-workings
+	 * of this handler to ensure it is not accessible from outside of this
+	 * function. Why? This is a stopgap implementation to provide the mechanisms
+	 * for webfonts to be defined in a theme's `theme.json` file for style
+	 * variations.
+	 *
+	 * @link  https://github.com/WordPress/gutenberg/issues/40472
+	 *
+	 * @since 6.0.0
+	 */
+	function _wp_theme_json_webfonts_handler() {
+		// Webfonts to be processed.
+		$registered_webfonts = array();
+
+		/**
+		 * Gets the webfonts from theme.json.
+		 *
+		 * @since 6.0.0
+		 *
+		 * @return array Array of defined webfonts.
+		 */
+		$fn_get_webfonts_from_theme_json = static function() {
+			// Get settings from theme.json.
+			$settings = WP_Theme_JSON_Resolver_6_0::get_merged_data()->get_settings();
+
+			// If in the editor, add webfonts defined in variations.
+			if ( is_admin() || ( defined( 'REST_REQUEST' ) && REST_REQUEST ) ) {
+				$variations = WP_Theme_JSON_Resolver_6_0::get_style_variations();
+				foreach ( $variations as $variation ) {
+					// Skip if fontFamilies are not defined in the variation.
+					if (
+						empty( $variation['settings'] ) ||
+						empty( $variation['settings']['typography'] ) ||
+						empty( $variation['settings']['typography']['fontFamilies'] )
+					) {
+						continue;
+					}
+
+					// Initialize the array structure.
+					if ( empty( $settings['typography'] ) ) {
+						$settings['typography'] = array();
+					}
+					if ( empty( $settings['typography']['fontFamilies'] ) ) {
+						$settings['typography']['fontFamilies'] = array();
+					}
+					if ( empty( $settings['typography']['fontFamilies']['theme'] ) ) {
+						$settings['typography']['fontFamilies']['theme'] = array();
+					}
+
+					// Combine variations with settings. Remove duplicates.
+					$settings['typography']['fontFamilies']['theme'] = array_merge( $settings['typography']['fontFamilies']['theme'], $variation['settings']['typography']['fontFamilies']['theme'] );
+					$settings['typography']['fontFamilies']          = array_unique( $settings['typography']['fontFamilies'] );
+				}
+			}
+
+			// Bail out early if there are no settings for webfonts.
+			if ( empty( $settings['typography'] ) || empty( $settings['typography']['fontFamilies'] ) ) {
```

Comment:
Font families can be defined at the top-level, as in:

```json
{
  "settings": {
      "typography": {
          "fontFamilies": []
      }
  }
}
```

But also can be defined per block, as in:

```json
{
  "settings": {
      "typography": {
          "fontFamilies": [] // Fonts for all.
      },
      "blocks": {
          "some/block": {
              "typography": {
                  "fontFamilies": [] // Unless a specific block defines its own.
              }
          }
      }
  }
}
```

Is this problematic for webfonts? How can a webfont be defined for a single block?

Reply:
>Is this problematic for webfonts? How can a webfont be defined for a single block?

@oandregal Top-level only in this first version. 

I remember a discussion thread (not sure where it happened) where @mtias indicated not having block-level web font selection. Also see https://github.com/WordPress/gutenberg/issues/40362.